qiita_db/metadata_template/base_metadata_template.py

```diff
@@ -1094,6 +1095,77 @@ def categories(self):
 
         return cols
 
+    def update(self, md_template):
+        r"""Update values in the sample template
+
+        Parameters
+        ----------
+        md_template : DataFrame
+            The metadata template file contents indexed by samples Ids
+
+        Raises
+        ------
+        QiitaDBError
+            If md_template and db do not have the same sample ids
+            If md_template and db do not have the same column headers
+            If self.can_be_updated is not True
+        """
+        conn_handler = SQLConnectionHandler()
+
+        # Clean and validate the metadata template given
+        new_map = self._clean_validate_template(md_template, self.study_id,
+                                                self.columns_restrictions)
+        # Retrieving current metadata
+        current_map = self._transform_to_dict(conn_handler.execute_fetchall(
+            "SELECT * FROM qiita.{0} WHERE {1}=%s".format(self._table,
+                                                          self._id_column),
+            (self.id,)))
+        dyn_vals = self._transform_to_dict(conn_handler.execute_fetchall(
+            "SELECT * FROM qiita.{0}".format(self._table_name(self.id))))
+
+        for k in current_map:
+            current_map[k].update(dyn_vals[k])
+            current_map[k].pop('study_id', None)
```

Comment:
Why not just bulk remove it from the subsequently produced DataFrame? It
will be more expressive, and is likely a more optimized operation.

On Mon, Jun 15, 2015 at 10:07 AM, josenavas <notifications@github.com>
wrote:

> In qiita_db/metadata_template/base_metadata_template.py
> <https://github.com/biocore/qiita/pull/1242#discussion_r32436795>:
>
> > +        conn_handler = SQLConnectionHandler()
> > +
> > +        # Clean and validate the metadata template given
> > +        new_map = self._clean_validate_template(md_template, self.study_id,
> > +                                                self.columns_restrictions)
> > +        # Retrieving current metadata
> > +        current_map = self._transform_to_dict(conn_handler.execute_fetchall(
> > +            "SELECT * FROM qiita.{0} WHERE {1}=%s".format(self._table,
> > +                                                          self._id_column),
> > +            (self.id,)))
> > +        dyn_vals = self._transform_to_dict(conn_handler.execute_fetchall(
> > +            "SELECT * FROM qiita.{0}".format(self._table_name(self.id))))
> > +
> > +        for k in current_map:
> > +            current_map[k].update(dyn_vals[k])
> > +            current_map[k].pop('study_id', None)
>
> I think I should comment here.
>
> Note that current_map is a dict of {sample: {metadata: value}} so in each
> loop we have to remove the 'study_id' entry.
>
> Re: situation where study_id won't be present: in the prep template. It
> is always present in the sample template but not in the prep templates. By
> doing current_map[k].pop('study_id', None) we make this code applicable
> to both templates.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/qiita/pull/1242/files#r32436795>.
>